src/NuGet.Core/NuGet.Commands/PackagesLockFileBuilder.cs

```diff
@@ -70,9 +71,14 @@ public PackagesLockFile CreateNuGetLockFile(LockFile assetsFile)
 
                 foreach (var projectReference in libraries.Where(e => e.Type == LibraryType.Project || e.Type == LibraryType.ExternalProject))
                 {
+                    var projectLibrary = assetsFile.Libraries.Single(a => a.Name == projectReference.Name && a.Version == projectReference.Version);
```

Comment:
How confident are you that this `Single(…)` call won't throw in some uncommon case?  It might be worth reviewing any internal guarantees on what `Libraries` can or cannot contain.

Also, is this now an O(n^2) algorithm?  If yes, what scenarios will it impact (e.g.:  creating a lock file for a project with many package references)?

Reply:
I'm not certain that it can't fail from duplicate matches, but I couldn't find an example when it happens. The input to this method is an assets file, and the libraries should have already been de-duplicated. Additionally, since this is a project reference, it shouldn't be possible to have multiple versions, as that would suggest that the project was evaluated more than once and had different versions. When I tested multiple projects with the same name, just different paths, NuGet crashed before getting to this method, so this code doesn't appear to be making it worse.

You were right about the algorithmic complexity, and I just pushed a fix.